File: C++/Edwin Villarraga/Grafos y Arboles/APSP+WAY.cpp
#include<bits/stdc++.h>
#include<cstdlib>
//--------------//
#define INF 9999
#define MAX 256
using namespace std;
const int V=4;
int next[V][V];
void printPath(vector<int> path) {
    int n = path.size()-1;
    for (int i = 0; i < n - 1; i++) {
        cout<<path[i]<<" -> ";
    }
    cout<<path[n-1]<<" -> "<<endl;
}
int constructPath(int u,int v){
    if(next[u][v]==-1){
        return 0;
    }
    vector<int>path;
    path.push_back(u);
    while(u != v){
        u=next[u][v];
        path.push_back(u);
    }
    printPath(path);
    return 1;
}
void printSolution(int dist[V][V]) {
       cout<<"La siguiente matriz muestra las distancias mas cortas entre cada par de vertices\n";
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][j] == INF) {
                    cout<<"INF \t";
                } else {
                    cout<<dist[i][j]<<"\t";
                }
            }
            cout<<endl;
        }
    }
void floydWarshall(int graph[V][V]) {
        int dist[V][V];
        int i, j, k;
        for (i = 0; i < V; i++) {
            for (j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
                if (graph[i][j] == INF) {
                    next[i][j] = -1;
                } else {
                    next[i][j] = j;
                }
            }
        }
        for (k = 0; k < V; k++) {
            for (i = 0; i < V; i++) {
                for (j = 0; j < V; j++) {
                    if (dist[i][k] == INF|| dist[k][j] == INF) {
                        continue;
                    }
                    if (dist[i][k] + dist[k][j] < dist[i][j]) {
                        dist[i][j] = dist[i][k] + dist[k][j];
                        next[i][j] = next[i][k];
                    }
                }
            }
        }
        printSolution(dist);
    }
int main() {
    int E,a,b,c;
    cin>>E;
    int graph[V][V];
    for(int i=0;i<V;i++){
        for(int j=0;j<V;j++){
            graph[i][j]=INF;
        }
    }
    for(int i=0;i<V;i++){
        graph[i][i]=0;
    }
    for(int i=0;i<E;i++){
        cin>>a>>b>>c;
        graph[a][b]=c;
    }
    floydWarshall(graph);
    constructPath(0,3);
}

File: Codigos estudiantes por lenguaje/Java/Edwin Villarraga/Grafos y Arboles/APSP+WAY.java
import java.util.*;

public class Main {

    static int INF = 9999;
    static int V = 0;
    static int[][] graph;
    static int[][] next;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        V = sc.nextInt();
        graph = new int[V][V];
        next = new int[V][V];
        int E = sc.nextInt();
        for (int i = 0; i < V; i++) {
            for (int j = 0; j < V; j++) {
                graph[i][j] = INF;
            }
        }
        for (int i = 0; i < V; i++) {
            graph[i][i] = 0;
        }
        for (int i = 0; i < E; i++) {
            graph[sc.nextInt()][sc.nextInt()] = sc.nextInt();
        }
        floydWarshall(graph);
        ArrayList<Integer> path;
        System.out.println("El camino mas corto desde 0 a 3: ");
        path = constructPath(0, 3);
        printPath(path);
    }

    static void floydWarshall(int graph[][]) {
        int dist[][] = new int[V][V];
        int i, j, k;
        for (i = 0; i < V; i++) {
            for (j = 0; j < V; j++) {
                dist[i][j] = graph[i][j];
                if (graph[i][j] == INF) {
                    next[i][j] = -1;
                } else {
                    next[i][j] = j;
                }
            }
        }
        for (k = 0; k < V; k++) {
            for (i = 0; i < V; i++) {
                for (j = 0; j < V; j++) {
                    if (dist[i][k] == INF|| dist[k][j] == INF) {
                        continue;
                    }
                    if (dist[i][j] > dist[i][k]+ dist[k][j]) {
                        dist[i][j] = dist[i][k]+ dist[k][j];
                        next[i][j] = next[i][k];
                    }
                }
            }
        }
        printSolution(dist);
    }

    static ArrayList<Integer> constructPath(int u, int v) {
        if (next[u][v] == -1) {
            return new ArrayList<>();
        }
        ArrayList<Integer> path = new ArrayList<>();
        path.add(u);
        while (u != v) {
            u = next[u][v];
            path.add(u);
        }
        return path;
    }

    static void printPath(ArrayList<Integer> path) {
        int n = path.size();
        for (int i = 0; i < n - 1; i++) {
            System.out.print(path.get(i) + " -> ");
        }
        System.out.println(path.get(n - 1) + " -> ");
    }

    static void printSolution(int dist[][]) {
        System.out.println("La siguiente matriz muestra las distancias "
                + "más cortas entre cada par de vertices");
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][j] == Main.INF) {
                    System.out.print("INF ");
                } else {
                    System.out.print(dist[i][j] + " ");
                }
            }
            System.out.println();
        }
    }
}
